crashtracker-ffi/src/collector_windows/api.rs

```diff
@@ -0,0 +1,737 @@
+                "{:x}",
+                native_frame.AddrPC.Offset - module.base_address
+            ));
+            frame.path.clone_from(&module.path);
+
+            if let Some(pdb_info) = &module.pdb_info {
+                frame.build_id = Some(format!("{:x}{:x}", pdb_info.signature, pdb_info.age));
+                frame.build_id_type = Some(datadog_crashtracker::BuildIdType::PDB);
+                frame.file_type = Some(datadog_crashtracker::FileType::PE);
+            }
+        }
+
+        stacktrace
+            .push_frame(frame, true)
+            .expect("Failed to add frame");
+    }
+
+    stacktrace.set_complete()?;
+    Ok(stacktrace)
+}
+
+struct ModuleInfo {
+    base_address: u64,
+    size: u64,
+    path: Option<String>,
+    pdb_info: Option<PdbInfo>,
+}
+
+struct PdbInfo {
+    age: u32,
+    signature: Guid,
+}
+
+unsafe fn list_modules(process_handle: HANDLE) -> anyhow::Result<Vec<ModuleInfo>> {
+    // Use EnumProcessModules to get a list of modules
+    let mut module_infos = Vec::new();
+
+    // Get the number of bytes required to store the array of module handles
+    let mut cb_needed = 0;
+    EnumProcessModules(process_handle, std::ptr::null_mut(), 0, &mut cb_needed)
+        .context("Failed to get module list size")?;
+
+    // Allocate enough space for the module handles
+    let modules_count = cb_needed as usize / size_of::<HMODULE>();
+    let mut hmodules: Vec<HMODULE> = Vec::with_capacity(modules_count);
+    let mut cb_actual = 0;
+
+    EnumProcessModules(
+        process_handle,
+        hmodules.as_mut_ptr(),
+        cb_needed,
+        &mut cb_actual,
+    )
+    .context("Failed to enumerate process modules")?;
+
+    hmodules.set_len(cb_actual as usize / size_of::<HMODULE>());
+
+    // Iterate through the module handles and retrieve information
+    for &hmodule in hmodules.iter() {
+        let mut module_info = MODULEINFO::default();
+        if GetModuleInformation(
+            process_handle,
+            hmodule,
+            &mut module_info,
+            size_of::<MODULEINFO>() as u32,
+        )
+        .is_err()
+        {
+            continue;
+        }
+
+        let module_path = get_module_path(process_handle, hmodule);
+
+        module_infos.push(ModuleInfo {
+            base_address: module_info.lpBaseOfDll as u64,
+            size: module_info.SizeOfImage as u64,
+            path: module_path.ok(),
+            pdb_info: get_pdb_info(process_handle, module_info.lpBaseOfDll as u64).ok(),
+        });
+    }
+
+    Ok(module_infos)
+}
+
+unsafe fn read_memory<T>(process_handle: HANDLE, address: u64) -> Result<T> {
+    let mut bytes_read = 0;
+    let mut value = MaybeUninit::<T>::uninit();
+    let size = size_of::<T>();
+
+    let result = ReadProcessMemory(
+        process_handle,
+        address as *const _,
+        value.as_mut_ptr() as *mut _,
+        size,
+        Some(&mut bytes_read),
+    );
+
+    if result.is_err() || bytes_read != size {
```

Comment:
anyhow::ensure

Reply:
Done in https://github.com/DataDog/libdatadog/pull/892/commits/ea8e7c31f38e2b3c86a21406ac52785ae29e7797